apps/hash-graph/lib/graph/src/snapshot.rs

```diff
@@ -0,0 +1,156 @@
+mod entity;
+mod metadata;
+mod ontology;
+
+use std::{fmt::Debug, pin::pin};
+
+use error_stack::{Context, Report, Result};
+use futures::{Sink, SinkExt, Stream, StreamExt, TryStreamExt};
+use serde::{Deserialize, Serialize};
+use type_system::{DataType, EntityType, PropertyType};
+
+pub use self::{
+    entity::EntityRecord,
+    metadata::{BlockProtocolModuleVersions, CustomGlobalMetadata},
+    ontology::{
+        CustomOntologyMetadata, OntologyTemporalMetadata, OntologyTypeMetadata, OntologyTypeRecord,
+    },
+};
+pub use crate::snapshot::metadata::SnapshotMetadata;
+use crate::{
+    knowledge::Entity,
+    store::{crud::Read, query::Filter, InsertionError, QueryError},
+};
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase", tag = "type")]
+pub enum SnapshotEntry {
+    Snapshot(SnapshotMetadata),
+    DataType(OntologyTypeRecord<DataType>),
+    PropertyType(OntologyTypeRecord<PropertyType>),
+    EntityType(OntologyTypeRecord<EntityType>),
+    Entity(EntityRecord),
```

Comment:
In a future PR do we perhaps want to rename these to `EntitySnapshotRecord` or something?

Just to try avoiding confusion with other kinds of 'records' we have within the system (especially when we have something like `RecordId`)

Reply:
Good point, I renamed it now: [Rename EntityRecord and OntologyTypeRecord](https://github.com/hashintel/hash/pull/2355/commits/5f56af83173caac42841f6d1f85aab42ae400844)
I find the name pretty long but it's better than having the confusion in place. If we replace the data we send to the backend, we may want to change it to `Entity`?